File: ts-reference/tests/example-program-gen/exp/accounts/State.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface StateFields {
  boolField: boolean
  u8Field: number
  i8Field: number
  u16Field: number
  i16Field: number
  u32Field: number
  i32Field: number
  f32Field: number
  u64Field: BN
  i64Field: BN
  f64Field: number
  u128Field: BN
  i128Field: BN
  bytesField: Array<number>
  stringField: string
  pubkeyField: PublicKey
  vecField: Array<BN>
  vecStructField: Array<types.FooStructFields>
  optionField: boolean | null
  optionStructField: types.FooStructFields | null
  structField: types.FooStructFields
  arrayField: Array<boolean>
  enumField1: types.FooEnumKind
  enumField2: types.FooEnumKind
  enumField3: types.FooEnumKind
  enumField4: types.FooEnumKind
}

export interface StateJSON {
  boolField: boolean
  u8Field: number
  i8Field: number
  u16Field: number
  i16Field: number
  u32Field: number
  i32Field: number
  f32Field: number
  u64Field: string
  i64Field: string
  f64Field: number
  u128Field: string
  i128Field: string
  bytesField: Array<number>
  stringField: string
  pubkeyField: string
  vecField: Array<string>
  vecStructField: Array<types.FooStructJSON>
  optionField: boolean | null
  optionStructField: types.FooStructJSON | null
  structField: types.FooStructJSON
  arrayField: Array<boolean>
  enumField1: types.FooEnumJSON
  enumField2: types.FooEnumJSON
  enumField3: types.FooEnumJSON
  enumField4: types.FooEnumJSON
}

export class State {
  readonly boolField: boolean
  readonly u8Field: number
  readonly i8Field: number
  readonly u16Field: number
  readonly i16Field: number
  readonly u32Field: number
  readonly i32Field: number
  readonly f32Field: number
  readonly u64Field: BN
  readonly i64Field: BN
  readonly f64Field: number
  readonly u128Field: BN
  readonly i128Field: BN
  readonly bytesField: Array<number>
  readonly stringField: string
  readonly pubkeyField: PublicKey
  readonly vecField: Array<BN>
  readonly vecStructField: Array<types.FooStruct>
  readonly optionField: boolean | null
  readonly optionStructField: types.FooStruct | null
  readonly structField: types.FooStruct
  readonly arrayField: Array<boolean>
  readonly enumField1: types.FooEnumKind
  readonly enumField2: types.FooEnumKind
  readonly enumField3: types.FooEnumKind
  readonly enumField4: types.FooEnumKind

  static readonly discriminator = Buffer.from([
    216, 146, 107, 94, 104, 75, 182, 177,
  ])

  static readonly layout = borsh.struct([
    borsh.bool("boolField"),
    borsh.u8("u8Field"),
    borsh.i8("i8Field"),
    borsh.u16("u16Field"),
    borsh.i16("i16Field"),
    borsh.u32("u32Field"),
    borsh.i32("i32Field"),
    borsh.f32("f32Field"),
    borsh.u64("u64Field"),
    borsh.i64("i64Field"),
    borsh.f64("f64Field"),
    borsh.u128("u128Field"),
    borsh.i128("i128Field"),
    borsh.vecU8("bytesField"),
    borsh.str("stringField"),
    borsh.publicKey("pubkeyField"),
    borsh.vec(borsh.u64(), "vecField"),
    borsh.vec(types.FooStruct.layout(), "vecStructField"),
    borsh.option(borsh.bool(), "optionField"),
    borsh.option(types.FooStruct.layout(), "optionStructField"),
    types.FooStruct.layout("structField"),
    borsh.array(borsh.bool(), 3, "arrayField"),
    types.FooEnum.layout("enumField1"),
    types.FooEnum.layout("enumField2"),
    types.FooEnum.layout("enumField3"),
    types.FooEnum.layout("enumField4"),
  ])

  constructor(fields: StateFields) {
    this.boolField = fields.boolField
    this.u8Field = fields.u8Field
    this.i8Field = fields.i8Field
    this.u16Field = fields.u16Field
    this.i16Field = fields.i16Field
    this.u32Field = fields.u32Field
    this.i32Field = fields.i32Field
    this.f32Field = fields.f32Field
    this.u64Field = fields.u64Field
    this.i64Field = fields.i64Field
    this.f64Field = fields.f64Field
    this.u128Field = fields.u128Field
    this.i128Field = fields.i128Field
    this.bytesField = fields.bytesField
    this.stringField = fields.stringField
    this.pubkeyField = fields.pubkeyField
    this.vecField = fields.vecField
    this.vecStructField = fields.vecStructField.map(
      (item) => new types.FooStruct({ ...item })
    )
    this.optionField = fields.optionField
    this.optionStructField =
      (fields.optionStructField &&
        new types.FooStruct({ ...fields.optionStructField })) ||
      null
    this.structField = new types.FooStruct({ ...fields.structField })
    this.arrayField = fields.arrayField
    this.enumField1 = fields.enumField1
    this.enumField2 = fields.enumField2
    this.enumField3 = fields.enumField3
    this.enumField4 = fields.enumField4
  }

  static async fetch(c: Connection, address: PublicKey): Promise<State | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(PROGRAM_ID)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[]
  ): Promise<Array<State | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(PROGRAM_ID)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): State {
    if (!data.slice(0, 8).equals(State.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = State.layout.decode(data.slice(8))

    return new State({
      boolField: dec.boolField,
      u8Field: dec.u8Field,
      i8Field: dec.i8Field,
      u16Field: dec.u16Field,
      i16Field: dec.i16Field,
      u32Field: dec.u32Field,
      i32Field: dec.i32Field,
      f32Field: dec.f32Field,
      u64Field: dec.u64Field,
      i64Field: dec.i64Field,
      f64Field: dec.f64Field,
      u128Field: dec.u128Field,
      i128Field: dec.i128Field,
      bytesField: Array.from(dec.bytesField),
      stringField: dec.stringField,
      pubkeyField: dec.pubkeyField,
      vecField: dec.vecField,
      vecStructField: dec.vecStructField.map((item) =>
        types.FooStruct.fromDecoded(item)
      ),
      optionField: dec.optionField,
      optionStructField:
        (dec.optionStructField &&
          types.FooStruct.fromDecoded(dec.optionStructField)) ||
        null,
      structField: types.FooStruct.fromDecoded(dec.structField),
      arrayField: dec.arrayField,
      enumField1: types.FooEnum.fromDecoded(dec.enumField1),
      enumField2: types.FooEnum.fromDecoded(dec.enumField2),
      enumField3: types.FooEnum.fromDecoded(dec.enumField3),
      enumField4: types.FooEnum.fromDecoded(dec.enumField4),
    })
  }

  toJSON(): StateJSON {
    return {
      boolField: this.boolField,
      u8Field: this.u8Field,
      i8Field: this.i8Field,
      u16Field: this.u16Field,
      i16Field: this.i16Field,
      u32Field: this.u32Field,
      i32Field: this.i32Field,
      f32Field: this.f32Field,
      u64Field: this.u64Field.toString(),
      i64Field: this.i64Field.toString(),
      f64Field: this.f64Field,
      u128Field: this.u128Field.toString(),
      i128Field: this.i128Field.toString(),
      bytesField: this.bytesField,
      stringField: this.stringField,
      pubkeyField: this.pubkeyField.toString(),
      vecField: this.vecField.map((item) => item.toString()),
      vecStructField: this.vecStructField.map((item) => item.toJSON()),
      optionField: this.optionField,
      optionStructField:
        (this.optionStructField && this.optionStructField.toJSON()) || null,
      structField: this.structField.toJSON(),
      arrayField: this.arrayField,
      enumField1: this.enumField1.toJSON(),
      enumField2: this.enumField2.toJSON(),
      enumField3: this.enumField3.toJSON(),
      enumField4: this.enumField4.toJSON(),
    }
  }

  static fromJSON(obj: StateJSON): State {
    return new State({
      boolField: obj.boolField,
      u8Field: obj.u8Field,
      i8Field: obj.i8Field,
      u16Field: obj.u16Field,
      i16Field: obj.i16Field,
      u32Field: obj.u32Field,
      i32Field: obj.i32Field,
      f32Field: obj.f32Field,
      u64Field: new BN(obj.u64Field),
      i64Field: new BN(obj.i64Field),
      f64Field: obj.f64Field,
      u128Field: new BN(obj.u128Field),
      i128Field: new BN(obj.i128Field),
      bytesField: obj.bytesField,
      stringField: obj.stringField,
      pubkeyField: new PublicKey(obj.pubkeyField),
      vecField: obj.vecField.map((item) => new BN(item)),
      vecStructField: obj.vecStructField.map((item) =>
        types.FooStruct.fromJSON(item)
      ),
      optionField: obj.optionField,
      optionStructField:
        (obj.optionStructField &&
          types.FooStruct.fromJSON(obj.optionStructField)) ||
        null,
      structField: types.FooStruct.fromJSON(obj.structField),
      arrayField: obj.arrayField,
      enumField1: types.FooEnum.fromJSON(obj.enumField1),
      enumField2: types.FooEnum.fromJSON(obj.enumField2),
      enumField3: types.FooEnum.fromJSON(obj.enumField3),
      enumField4: types.FooEnum.fromJSON(obj.enumField4),
    })
  }
}

File: tests/client_gen/example_program_gen/accounts/state.py
import typing
from dataclasses import dataclass
from construct import Construct
from solders.pubkey import Pubkey
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
import borsh_construct as borsh
from anchorpy.coder.accounts import ACCOUNT_DISCRIMINATOR_SIZE
from anchorpy.error import AccountInvalidDiscriminator
from anchorpy.utils.rpc import get_multiple_accounts
from anchorpy.borsh_extension import BorshPubkey
from ..program_id import PROGRAM_ID
from .. import types


class StateJSON(typing.TypedDict):
    bool_field: bool
    u8_field: int
    i8_field: int
    u16_field: int
    i16_field: int
    u32_field: int
    i32_field: int
    f32_field: float
    u64_field: int
    i64_field: int
    f64_field: float
    u128_field: int
    i128_field: int
    bytes_field: list[int]
    string_field: str
    pubkey_field: str
    vec_field: list[int]
    vec_struct_field: list[types.foo_struct.FooStructJSON]
    option_field: typing.Optional[bool]
    option_struct_field: typing.Optional[types.foo_struct.FooStructJSON]
    struct_field: types.foo_struct.FooStructJSON
    array_field: list[bool]
    enum_field1: types.foo_enum.FooEnumJSON
    enum_field2: types.foo_enum.FooEnumJSON
    enum_field3: types.foo_enum.FooEnumJSON
    enum_field4: types.foo_enum.FooEnumJSON


@dataclass
class State:
    discriminator: typing.ClassVar = b"\xd8\x92k^hK\xb6\xb1"
    layout: typing.ClassVar = borsh.CStruct(
        "bool_field" / borsh.Bool,
        "u8_field" / borsh.U8,
        "i8_field" / borsh.I8,
        "u16_field" / borsh.U16,
        "i16_field" / borsh.I16,
        "u32_field" / borsh.U32,
        "i32_field" / borsh.I32,
        "f32_field" / borsh.F32,
        "u64_field" / borsh.U64,
        "i64_field" / borsh.I64,
        "f64_field" / borsh.F64,
        "u128_field" / borsh.U128,
        "i128_field" / borsh.I128,
        "bytes_field" / borsh.Bytes,
        "string_field" / borsh.String,
        "pubkey_field" / BorshPubkey,
        "vec_field" / borsh.Vec(typing.cast(Construct, borsh.U64)),
        "vec_struct_field"
        / borsh.Vec(typing.cast(Construct, types.foo_struct.FooStruct.layout)),
        "option_field" / borsh.Option(borsh.Bool),
        "option_struct_field" / borsh.Option(types.foo_struct.FooStruct.layout),
        "struct_field" / types.foo_struct.FooStruct.layout,
        "array_field" / borsh.Bool[3],
        "enum_field1" / types.foo_enum.layout,
        "enum_field2" / types.foo_enum.layout,
        "enum_field3" / types.foo_enum.layout,
        "enum_field4" / types.foo_enum.layout,
    )
    bool_field: bool
    u8_field: int
    i8_field: int
    u16_field: int
    i16_field: int
    u32_field: int
    i32_field: int
    f32_field: float
    u64_field: int
    i64_field: int
    f64_field: float
    u128_field: int
    i128_field: int
    bytes_field: bytes
    string_field: str
    pubkey_field: Pubkey
    vec_field: list[int]
    vec_struct_field: list[types.foo_struct.FooStruct]
    option_field: typing.Optional[bool]
    option_struct_field: typing.Optional[types.foo_struct.FooStruct]
    struct_field: types.foo_struct.FooStruct
    array_field: list[bool]
    enum_field1: types.foo_enum.FooEnumKind
    enum_field2: types.foo_enum.FooEnumKind
    enum_field3: types.foo_enum.FooEnumKind
    enum_field4: types.foo_enum.FooEnumKind

    @classmethod
    async def fetch(
        cls,
        conn: AsyncClient,
        address: Pubkey,
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.Optional["State"]:
        resp = await conn.get_account_info(address, commitment=commitment)
        info = resp.value
        if info is None:
            return None
        if info.owner != program_id:
            raise ValueError("Account does not belong to this program")
        bytes_data = info.data
        return cls.decode(bytes_data)

    @classmethod
    async def fetch_multiple(
        cls,
        conn: AsyncClient,
        addresses: list[Pubkey],
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.List[typing.Optional["State"]]:
        infos = await get_multiple_accounts(conn, addresses, commitment=commitment)
        res: typing.List[typing.Optional["State"]] = []
        for info in infos:
            if info is None:
                res.append(None)
                continue
            if info.account.owner != program_id:
                raise ValueError("Account does not belong to this program")
            res.append(cls.decode(info.account.data))
        return res

    @classmethod
    def decode(cls, data: bytes) -> "State":
        if data[:ACCOUNT_DISCRIMINATOR_SIZE] != cls.discriminator:
            raise AccountInvalidDiscriminator(
                "The discriminator for this account is invalid"
            )
        dec = State.layout.parse(data[ACCOUNT_DISCRIMINATOR_SIZE:])
        return cls(
            bool_field=dec.bool_field,
            u8_field=dec.u8_field,
            i8_field=dec.i8_field,
            u16_field=dec.u16_field,
            i16_field=dec.i16_field,
            u32_field=dec.u32_field,
            i32_field=dec.i32_field,
            f32_field=dec.f32_field,
            u64_field=dec.u64_field,
            i64_field=dec.i64_field,
            f64_field=dec.f64_field,
            u128_field=dec.u128_field,
            i128_field=dec.i128_field,
            bytes_field=dec.bytes_field,
            string_field=dec.string_field,
            pubkey_field=dec.pubkey_field,
            vec_field=dec.vec_field,
            vec_struct_field=list(
                map(
                    lambda item: types.foo_struct.FooStruct.from_decoded(item),
                    dec.vec_struct_field,
                )
            ),
            option_field=dec.option_field,
            option_struct_field=(
                None
                if dec.option_struct_field is None
                else types.foo_struct.FooStruct.from_decoded(dec.option_struct_field)
            ),
            struct_field=types.foo_struct.FooStruct.from_decoded(dec.struct_field),
            array_field=dec.array_field,
            enum_field1=types.foo_enum.from_decoded(dec.enum_field1),
            enum_field2=types.foo_enum.from_decoded(dec.enum_field2),
            enum_field3=types.foo_enum.from_decoded(dec.enum_field3),
            enum_field4=types.foo_enum.from_decoded(dec.enum_field4),
        )

    def to_json(self) -> StateJSON:
        return {
            "bool_field": self.bool_field,
            "u8_field": self.u8_field,
            "i8_field": self.i8_field,
            "u16_field": self.u16_field,
            "i16_field": self.i16_field,
            "u32_field": self.u32_field,
            "i32_field": self.i32_field,
            "f32_field": self.f32_field,
            "u64_field": self.u64_field,
            "i64_field": self.i64_field,
            "f64_field": self.f64_field,
            "u128_field": self.u128_field,
            "i128_field": self.i128_field,
            "bytes_field": list(self.bytes_field),
            "string_field": self.string_field,
            "pubkey_field": str(self.pubkey_field),
            "vec_field": self.vec_field,
            "vec_struct_field": list(
                map(lambda item: item.to_json(), self.vec_struct_field)
            ),
            "option_field": self.option_field,
            "option_struct_field": (
                None
                if self.option_struct_field is None
                else self.option_struct_field.to_json()
            ),
            "struct_field": self.struct_field.to_json(),
            "array_field": self.array_field,
            "enum_field1": self.enum_field1.to_json(),
            "enum_field2": self.enum_field2.to_json(),
            "enum_field3": self.enum_field3.to_json(),
            "enum_field4": self.enum_field4.to_json(),
        }

    @classmethod
    def from_json(cls, obj: StateJSON) -> "State":
        return cls(
            bool_field=obj["bool_field"],
            u8_field=obj["u8_field"],
            i8_field=obj["i8_field"],
            u16_field=obj["u16_field"],
            i16_field=obj["i16_field"],
            u32_field=obj["u32_field"],
            i32_field=obj["i32_field"],
            f32_field=obj["f32_field"],
            u64_field=obj["u64_field"],
            i64_field=obj["i64_field"],
            f64_field=obj["f64_field"],
            u128_field=obj["u128_field"],
            i128_field=obj["i128_field"],
            bytes_field=bytes(obj["bytes_field"]),
            string_field=obj["string_field"],
            pubkey_field=Pubkey.from_string(obj["pubkey_field"]),
            vec_field=obj["vec_field"],
            vec_struct_field=list(
                map(
                    lambda item: types.foo_struct.FooStruct.from_json(item),
                    obj["vec_struct_field"],
                )
            ),
            option_field=obj["option_field"],
            option_struct_field=(
                None
                if obj["option_struct_field"] is None
                else types.foo_struct.FooStruct.from_json(obj["option_struct_field"])
            ),
            struct_field=types.foo_struct.FooStruct.from_json(obj["struct_field"]),
            array_field=obj["array_field"],
            enum_field1=types.foo_enum.from_json(obj["enum_field1"]),
            enum_field2=types.foo_enum.from_json(obj["enum_field2"]),
            enum_field3=types.foo_enum.from_json(obj["enum_field3"]),
            enum_field4=types.foo_enum.from_json(obj["enum_field4"]),
        )
